### PR TITLE
Add support for logitech TAP devices

### PR DESCRIPTION
**Pull request by vcdmp**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c [hunk 1]
```diff
@@ -0,0 +1,860 @@
...
+fu_logitech_tap_touch_device_write_enable(FuLogitechTapTouchDevice *self,
+					  gboolean in_ap,
+					  gboolean write_ap,
+					  guint32 end,
+					  guint32 checksum,
+					  GError **error)
+{
+	guint8 set_data[TOUCH_HID_SET_DATA_LEN];
+	guint8 delay;
+	memset(set_data, 0, TOUCH_HID_SET_DATA_LEN);
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-touch-device.c:157]: just do `guint8 set_data[TOUCH_HID_SET_DATA_LEN] = {0};` pla

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c [hunk 2]
```diff
@@ -0,0 +1,860 @@
...
+	memset(get_data, 0, TOUCH_HID_GET_DATA_LEN);
+	set_data[0] = 0x03;
+	set_data[1] = 0xA3;
+	set_data[2] = 0x01;
+	set_data[3] = 0x01;
+	set_data[4] = 0x80;
+
+	g_debug("SRS:TODO TEMP fu_logitech_tap_touch_device_check_busy Entering");
+	while (max_retry--) {
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-touch-device.c:210]: needs to use `fu_device_retry()`

File: plugins/logitech-tap/fu-logitech-tap-touch-device.h [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright 1999-2023 Logitech, Inc.
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-touch-device.h:2]: just `Copyright 2024 Logitech, Inc.` please

**Comment by vcdmp** [comment]
Richard,  This feature is still under development. I will refactor/cleanup, take care of any review feedback etc. 

Currently we are consistently observing "ioctl error: Broken pipe [32]" in bootloader mode.
Raised the issue here: https://groups.google.com/g/fwupd/c/0_0H1RIH8yY

First time leveraging attach/detach features, I am in uncharted territory here, looking forward to your help.

**Comment by hughsie** [comment]
> we are consistently observing "ioctl error: Broken pipe [32]" in bootloader mode

Is the device rebooting without returning from the ioctl perhaps? if so, we can use error_local and g_error_matches to ignore that specific code.

**Comment by vcdmp** [comment]
> > we are consistently observing "ioctl error: Broken pipe [32]" in bootloader mode
> 
> Is the device rebooting without returning from the ioctl perhaps? if so, we can use error_local and g_error_matches to ignore that specific code.

Here is what Vendor said:
"FW would send Interrupt-In packet to response most of commands, as the same way to send touch point report to system. Please use read() for both AP/BL mode."

I see that there are fu_hid_device_set_report/fu_usb_device_interrupt_transfer API, but they are for USB. How do we achieve the same for HIDRAW device?

**Comment by vcdmp** [comment]
[fwupdtool_update_4005_to_4007.zip](https://github.com/user-attachments/files/16820610/fwupdtool_update_4005_to_4007.zip)

**Comment by hughsie** [comment]
@vcdmp i've rebased this and fixed up a ton of issues -- I hope i've not broken anything too badly. There are some more fixmes, but I'm out of time for today. If you can answer any of the questions in the source that'd be great :)

**Comment by vcdmp** [comment]
> @vcdmp i've rebased this and fixed up a ton of issues -- I hope i've not broken anything too badly. There are some more fixmes, but I'm out of time for today. If you can answer any of the questions in the source that'd be great :)

I was also working on refactoring/cleanup yesterday and was about to push my changes. I will merge our changes and test. It is broken, but not that bad. 

As for "FIXME use FuChunk", since this is relatively small image size, loading it fully. 
1. We have to look for magic tag located between certain offset, so easy to leverage fu_memmem_safe(). 
2. We also have to push image blocks located between certain offsets
How do we achieve above using chunks. Any pointers!

Let me dig more for FIXME in .rs file.

TAP is composite device and now two subdevices are upgradeable (Touch/HDMI). Both are independent of each other and can be updated in parallel. But after upgrade HDMI reboots the composite device, so prefer to update HDMI last. 
Options are:
1. Create custom header and embed both HDMI/Touch images
2. Let platform (ChromeOS), create multiple UDEV rules and invoke updater in a specific order, with specific image
I am not in favor of option #1 here (also clueless). What is your vote? Any pointers?

**Comment by hughsie** [comment]
> I was also working on refactoring/cleanup yesterday and was about to push my changes.

OOpa, sorry.

> As for "FIXME use FuChunk", since this is relatively small image size, loading it fully.

Ohh this was more for readability than memory use.

>    How do we achieve above using chunks. Any pointers!

Sure, I can look at this again next week.

> Let me dig more for FIXME in .rs file.

Thanks.

> TAP is composite device and now two subdevices are upgradeable

>    Create custom header and embed both HDMI/Touch images

Eww.

> Let platform (ChromeOS), create multiple UDEV rules and invoke updater in a specific order

I don't think the udev loader solution is going to work medium term. I think the plan now is to 100% move to the daemon due to issues like this.

> Any pointers?

Could we make one the parent of the other, and use FU_DEVICE_PRIVATE_FLAG_INSTALL_PARENT_FIRST to influence the order they get installed?

**Comment by vcdmp** [comment]
@hughsie  
Richard, some recent changes has caused regression with this device (video4linux subsystem).

Every thing works when we use 1.9.24 codebase. But same device on same platform/setup fails with latest codebase.

With latest codebase, device is seen (PID=0876), but not processed for following reason:
"removing #### from backend cache as no possible plugin"

Verbose logs attached for both working/non working usecases.
[device_not_recognized.zip](https://github.com/user-attachments/files/16937399/device_not_recognized.zip)

Reaching out, as I cannot narrow down which fix cause regression.

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+	fu_struct_logitech_tap_touch_set_payload_len(st, 0x01);
+
+	if (crc)
+		fu_struct_logitech_tap_touch_set_response_len(st, datasz);
+	else
+		fu_struct_logitech_tap_touch_set_response_len(st, 0x0);
+	fu_byte_array_append_uint8(st, FU_STRUCT_LOGITECH_TAP_TOUCH_HID_CMD_GET_AP_CRC);
+
+	/* hid report to query crc info of dataflash/pflash (DF/AP) block */
+	if (crc) {
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-touch-device.c:247]: `!= NULL` please -- then it's obvious it's a pointer and not being compared against non-zero

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c [hunk 5]
```diff
@@ -0,0 +1,913 @@
...
+						       FALSE,
+						       in_ap ? TRUE : FALSE,
+						       end + 1,
+						       firmware_checksum,
+						       error))
+		return FALSE;
+	fu_device_sleep(FU_DEVICE(self), 10);
+
+	// FIXME use FuChunk
+	for (guint32 i = start; i <= end; i += TAP_TOUCH_TRANSFER_BLOCK_SIZE) {
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-touch-device.c:597]: Still this; I think FuChunk will clean up a lot of this

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c [hunk 6]
```diff
@@ -0,0 +1,913 @@
...
+	locker =
+	    fu_device_locker_new_full(FU_DEVICE(self),
+				      (FuDeviceLockerFunc)fu_logitech_tap_touch_device_enable_tde,
+				      (FuDeviceLockerFunc)fu_logitech_tap_touch_device_disable_tde,
+				      error);
+	if (locker == NULL)
+		return FALSE;
+	/* recommended to retry at least once */
+	do {
+		fu_progress_step_done(progress);
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-touch-device.c:731]: I think all of this belongs in a `fu_device_retry` helper --it'll also unindent the code a bit too.

**Comment by vcdmp** [comment]
> @vcdmp
> 
> > UpdateError: No vendor ID set
> 
> Were you expecting to see a vendor ID of `USB:0x222A` there?

Yes, it is a hidraw. Also for PIR Sensor sub-device (hidraw) we expect Logitech PID: 046D

**Comment by vcdmp** [comment]
@hughsie 
 Hi Richard, Any updates with the regression issue? 
 Suspecting some issue with descriptors. This is composite device consist of hidraw/hub/video4linux etc. 
 Seeing quite a few errors in non-working log, they don't appear in working log (1.9.24 codebase)
 
22:19:03.347 FuCommon             failed to read as mapped file, so reading /sys/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1.2/descriptors of size 0x1a3: Failed to map /sys/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1.2/descriptors' /sys/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1.2/descriptors': mmap() failed: No such device

Eventually it removes device:
removing /sys/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1.2 from backend cache as no possible plugin

**Comment by vcdmp** [comment]
@hughsie 
Hey Richard, just a gentle reminder to look into regression issue. I am blocked and cannot proceed, because  device is not recognized.

**Comment by vcdmp** [comment]
I am not blocked anymore, working on your code earlier review feedback.

**Comment by vcdmp** [comment]
Did more cleanup/refactoring. Still seeing following regression issues:
 
 PIR sensor (PID:0x0872) device entry appears twice, once with Serial and one without Serial. Both has same GUID but different DeviceId
 
 HDMI (video4linux, PID:0x0872), is still not seen by daemon. adding vendor_id manually didn't help either.

**Comment by vcdmp** [comment]
Upgrade logs attached for clarity
[fwupd_fwupdmgr_vv.zip](https://github.com/user-attachments/files/17052437/fwupd_fwupdmgr_vv.zip)

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c (hunk 4 above)
Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap-touch-device.c]: Fixed

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c (hunk 5 above)
Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap-touch-device.c]: Fixed

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c (hunk 6 above)
Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap-touch-device.c]: Fixed

File: plugins/logitech-tap/fu-logitech-tap-touch-device.c [hunk 7]
```diff
@@ -0,0 +1,913 @@
...
+
+		/* validate ic name and protocol version specified in firmware file */
+		// move to fu-logitech-tap-touch-firmware.c
```
[the remark this answers is not in the document]
Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap-touch-device.c]: Fixed

**Comment by hughsie** [comment]
@vcdmp I've pushed two fixups, the latter may have broken the crc calculation. Coverity would *not* have liked those random offsets, and I think a stream is probably way more efficient than a buffer copy.

**Comment by hughsie** [comment]
@vcdmp while we're in this plugin, any chance of converting the FuLogitechTapSensorDevice constants like kHidMcuTde into rustgen enums and things like:

	guint8 buf[HID_SET_DATA_LEN] = {
	    kHidMcuTdeReportId,
	    kHidMcuTdeModeSelector,
	    kHidMcuTdeModeEnable,
	    0,
	    0,
	};

into a rustgen struct? I think it would help readability (and debugability!) a lot.

**Comment by vcdmp** [comment]
> @vcdmp I've pushed two fixups, the latter may have broken the crc calculation. Coverity would _not_ have liked those random offsets, and I think a stream is probably way more efficient than a buffer copy.

stream didn't worked. Inverting (~) didn't helped either. Those offsets are validated before use, so we don't expect runtime issues. Any other pointer? Or restore original code for now?

**Comment by vcdmp** [comment]
> > @vcdmp I've pushed two fixups, the latter may have broken the crc calculation. Coverity would _not_ have liked those random offsets, and I think a stream is probably way more efficient than a buffer copy.
> 
> stream didn't worked. Inverting (~) didn't helped either. Those offsets are validated before use, so we don't expect runtime issues. Any other pointer? Or restore original code for now?

I have reverted your crc changes for now. We can revisit later if needed.

**Comment by vcdmp** [comment]
> @vcdmp while we're in this plugin, any chance of converting the FuLogitechTapSensorDevice constants like kHidMcuTde into rustgen enums and things like:
> 
> ```
> guint8 buf[HID_SET_DATA_LEN] = {
>     kHidMcuTdeReportId,
>     kHidMcuTdeModeSelector,
>     kHidMcuTdeModeEnable,
>     0,
>     0,
> };
> ```
> 
> into a rustgen struct? I think it would help readability (and debugability!) a lot.

Because of regression issue with HDMI subdevice is not getting recognized, I cannot test it. I have created separate [draft PR](https://github.com/fwupd/fwupd/pull/7757) to address this

All logs are attached to this PR, need your help with regression issues, again everything on same setup works, with last stable release 1.9.24 :
https://github.com/fwupd/fwupd/pull/7644#issuecomment-2357496138
https://github.com/fwupd/fwupd/pull/7644#issuecomment-2339294362

**Comment by vcdmp** [comment]
@hughsie 
Removed adding vendor id manually. https://github.com/fwupd/fwupd/pull/7762 fix helps

We still have regression issue with HDMI (video4linux) device, as documented here:
https://github.com/fwupd/fwupd/pull/7644#issuecomment-2357496138
https://github.com/fwupd/fwupd/pull/7644#issuecomment-2339294362

Possible to review/approve this PR! Because HDMI regression is not relevant to this PR.

**Comment by hughsie** [comment]
> PIR sensor (PID:0x0872) device entry appears twice, once with Serial and one without Serial. Both has same GUID but different DeviceId

Right, the daemon is finding `444af8ce-dcd7-5b46-9c5f-13a59a5a5b8a ← USB\VID_046D&PID_0872` (which we're ignoring, i.e. the `removing /sys/devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4.1 from backend cache as no possible plugin`) and `35493092-8e6c-54f1-bba4-2e01f3fcc4bf ← HIDRAW\VEN_046D&DEV_0872` which we're probing (`emit added from logitech_tap: da3b508a1f35c3abd1d12f647c4080c340dcb693`).

I think the code is doing the right thing here.

> HDMI sub-device (video4linux, PID:0x0872), is still not seen by daemon. adding vendor_id manually didn't help either.

I broke this, give me an hour or so.

**Comment by hughsie** [comment]
@vcdmp I think https://github.com/fwupd/fwupd/pull/7765 should fix this. Please let me know.

**Comment by vcdmp** [comment]
> @vcdmp I think #7765 should fix this. Please let me know.

Do I need to pull any other changes from another branch? Tested with 'hughsie/FuV4lDevice', not helping. vv logs attached
[fwupdtool.zip](https://github.com/user-attachments/files/17078567/fwupdtool.zip)

**Comment by hughsie** [comment]
@vcdmp thanks for the quick testing - I see:

    17:14:57.154 FuEngine             logitech_tap ignoring: failed to setup: only video capture device are supported

I've pushed ecf6039a417967eb2f1a78e2570b9dd0bc7c87e1 -- can you cherry pick that and grab the fwupdtool log again please.

**Comment by vcdmp** [comment]
> @vcdmp thanks for the quick testing - I see:
> 
> ```
> 17:14:57.154 FuEngine             logitech_tap ignoring: failed to setup: only video capture device are supported
> ```
> 
> I've pushed [ecf6039](https://github.com/fwupd/fwupd/commit/ecf6039a417967eb2f1a78e2570b9dd0bc7c87e1) -- can you cherry pick that and grab the fwupdtool log again please.

Yahoo.... That helped.  New vv logs attached. 
FYI, I have v4l device, VS code all necessary setup ready, let me know if you want live troubleshooting, step thru probe(), add troubleshooting logs etc.

[fwupdtool.zip](https://github.com/user-attachments/files/17078954/fwupdtool.zip)

**Comment by hughsie** [comment]
> Yahoo.... That helped. New vv logs attached.

That looks great, thanks. I've squashed down the patches in https://github.com/fwupd/fwupd/pull/7765 and your review approval would be lovely. Have a great weekend.

**Comment by vcdmp** [comment]
> > Yahoo.... That helped. New vv logs attached.
> 
> That looks great, thanks. I've squashed down the patches in #7765 and your review approval would be lovely. Have a great weekend.

Tested with multiple v4l devices connected (Scribe+Tap), not seeing any errors. Would you please review/approve this PR. Then I will truly enjoy week-end :-)

**Comment by vcdmp** [comment]
@hughsie @superm1 Gentle reminder to review/approve this PR

**Comment by vcdmp** [comment]
Tested with latest main branch commits. Not seeing any issues. vv logs attached for reference.

Please review/approve this PR

[fwupd_fwuptool_vv.zip](https://github.com/user-attachments/files/17102795/fwupd_fwuptool_vv.zip)

**Comment by hughsie** [comment]
I've got this on the list to review for tomorrow.

**Comment by vcdmp** [comment]
> I've got this on the list to review for tomorrow.

Thanks in advance @hughsie. Let us know if anything is needed from Logitech side.

**Comment by hughsie** [comment]
@vcdmp I've opened https://github.com/fwupd/fwupd/pull/7778 and https://github.com/fwupd/fwupd/pull/7779 for new functions we can use here. I've cherry picked both commits onto your branch and fixed up FuLogitechTapTouchFirmware to use them.

I did the input stream helpers for two reasons -- Coverity got really unhappy about using the untrusted offsets as read offsets to other places (if we fuzzed the loader, it would have exploded), and we also want to keep the RSS as low as possible on ChromeOS as some of the images are pretty large and we can't always just page them into a buffer in one big chunk. Can you verify I've not broken anything again please.

**Comment by vcdmp** [comment]
> @vcdmp I've opened #7778 and #7779 for new functions we can use here. I've cherry picked both commits onto your branch and fixed up FuLogitechTapTouchFirmware to use them.
> 
> I did the input stream helpers for two reasons -- Coverity got really unhappy about using the untrusted offsets as read offsets to other places (if we fuzzed the loader, it would have exploded), and we also want to keep the RSS as low as possible on ChromeOS as some of the images are pretty large and we can't always just page them into a buffer in one big chunk. Can you verify I've not broken anything again please.

Seeing following error:
failed write-firmware 'base stream was 0xf3a0 bytes in size, and tried to create partial stream @0x2000 of 0xea00 bytes'

Again, I have v4l device, VS code, all necessary setup ready. Let me know if you want live troubleshooting, step thru code, add troubleshooting logs etc.

**Comment by hughsie** [comment]
>  failed write-firmware 'base stream was 0xf3a0 bytes in size, and tried to create partial stream @0x2000 of 0xea00 bytes'

Okay, so that's telling us there's a real bug! The old code would have accessed memory out of bounds and caused valgrind/asan to explode. If the base stream (the firmware binary) is 0xf3a0 in size and we start reading at offset 0x2000 the biggest slice we can make is 0xD3A0 and we're asking for a slice of 0xEA00 i.e. 5728 bytes more than we have. I don't know what the slice size is really supposed to be I'm afraid, but I think the plugin code might be wrong.

**Comment by vcdmp** [comment]
> > failed write-firmware 'base stream was 0xf3a0 bytes in size, and tried to create partial stream @0x2000 of 0xea00 bytes'
> 
> Okay, so that's telling us there's a real bug! The old code would have accessed memory out of bounds and caused valgrind/asan to explode. If the base stream (the firmware binary) is 0xf3a0 in size and we start reading at offset 0x2000 the biggest slice we can make is 0xD3A0 and we're asking for a slice of 0xEA00 i.e. 5728 bytes more than we have. I don't know what the slice size is really supposed to be I'm afraid, but I think the plugin code might be wrong.

Yes, troubleshooting. Need to readjust start/offset/end sizes, because now we are dealing with partial buffers/stream

**Comment by hughsie** [comment]
@vcdmp I've pushed a fix and also simplified the firmware splitting -- can you check again please. If that works then I'm happy.

**Comment by vcdmp** [comment]
Thanks @hughsie for all your help/support. I have also refactored some of the block size/offset logic, will merge/test.

Problem here is (fu_input_stream_compute_crc16), it is not working at all, I am troubleshooting.

**Comment by vcdmp** [comment]
@hughsie @superm1 Fixed few more issues. Please review. vv logs attached
[fwupdtool.zip](https://github.com/user-attachments/files/17140604/fwupdtool.zip)

**Comment by hughsie** [comment]
Rebased and squashed. Good job as always @vcdmp

**Review by hughsie — APPROVED** [review]

**Pull request merged**